File: app/javascript/cyvasse/units.js
// The eleven Cyvasse unit types and the nineteen-piece army.
//
// Ported from the legacy factory, amcritchie/Cyvasse
// app/assets/javascripts/LoadFactory/createUnits.js. Where the Ruby models
// (app/models/units/*.rb) disagree with the factory, the factory wins: it is
// what the browser actually played with. The Ruby files carried display
// strings ("2 + 2", "Moves in a straight line") and a stale flank column; the
// numbers below are the ones every legacy move was computed from.
//
// `rank` keeps the legacy spelling of the vanguard ("vangaurd") out of the
// data: nothing ever branched on it, so it is written correctly here.

export const UNIT_TYPES = Object.freeze({
  rabble: unit("rabble", "Rabble", "vanguard", { attack: 1, defence: 1, moveRange: 3, attackRange: 0, flank: 2, trump: [] }),
  spearman: unit("spearman", "Spearman", "vanguard", { attack: 2, defence: 2, moveRange: 2, attackRange: 0, flank: 1, trump: ["lighthorse"] }),
  elephant: unit("elephant", "Elephant", "vanguard", { attack: 4, defence: 4, moveRange: 3, attackRange: 0, flank: 1, trump: [] }),
  lighthorse: unit("lighthorse", "Light Horse", "cavalry", { attack: 2, defence: 2, moveRange: 3, attackRange: 0, flank: 1, trump: [] }),
  heavyhorse: unit("heavyhorse", "Heavy Horse", "cavalry", { attack: 3, defence: 3, moveRange: 2, attackRange: 0, flank: 1, trump: [] }),
  crossbowman: unit("crossbowman", "Crossbowman", "range", { attack: 2, defence: 1, moveRange: 1, attackRange: 2, flank: 0, trump: ["elephant"] }),
  trebuchet: unit("trebuchet", "Trebuchet", "range", { attack: 1, defence: 1, moveRange: 0, attackRange: 3, flank: 0, trump: ["dragon"] }),
  catapult: unit("catapult", "Catapult", "range", { attack: 3, defence: 1, moveRange: 2, attackRange: 3, flank: 0, trump: ["dragon"] }),
  dragon: unit("dragon", "Dragon", "unique", { attack: 5, defence: 5, moveRange: 10, attackRange: 0, flank: 0, trump: [] }),
  king: unit("king", "King", "unique", { attack: 2, defence: 2, moveRange: 2, attackRange: 0, flank: 0, trump: [] }),
  mountain: unit("mountain", "Mountain", "mountain", { attack: 9, defence: 9, moveRange: 0, attackRange: 0, flank: 0, trump: [] })
});

// The army, by the legacy data-index (1-based). Every stored setup string
// ("11:39|6:38|...") names units by this index, so the order is load-bearing.
export const ARMY = Object.freeze([
  "rabble", "rabble", "rabble",
  "spearman", "spearman",
  "elephant", "elephant",
  "lighthorse", "lighthorse",
  "heavyhorse", "heavyhorse",
  "crossbowman", "crossbowman",
  "trebuchet",
  "catapult",
  "dragon",
  "king",
  "mountain", "mountain"
]);

export const ARMY_SIZE = ARMY.length;

// The type of the unit at a 1-based army index.
export function typeAt(index) {
  const type = ARMY[index - 1];
  if (!type) throw new RangeError(`no unit at army index ${index}`);
  return UNIT_TYPES[type];
}

function unit(codename, name, rank, stats) {
  return Object.freeze({ codename, name, rank, ...stats, trump: Object.freeze(stats.trump) });
}

File: app/models/cyvasse_rules/units.rb
module CyvasseRules
  # The eleven unit types and the nineteen-piece army: a Ruby mirror of
  # app/javascript/cyvasse/units.js, which is the source of truth. The server
  # validates every online move with these numbers, so they must equal the
  # browser's: test/models/cyvasse_rules/js_agreement_test.rb replays the JS
  # engine's own answers (test/fixtures/files/rules_agreement.json) against
  # this port, and test/lib/engine_rulebook_agreement_test.rb pins the table.
  module Units
    Type = Data.define(:codename, :name, :rank, :attack, :defence, :move_range, :attack_range, :trump) do
      def mountain? = codename == "mountain"
      def dragon? = codename == "dragon"
      def king? = codename == "king"
      def cavalry? = rank == "cavalry"
      def range? = rank == "range"
    end

    def self.type(codename, name, rank, attack:, defence:, move:, range:, trump: [])
      Type.new(codename:, name:, rank:, attack:, defence:, move_range: move, attack_range: range, trump: trump.freeze)
    end
    private_class_method :type

    TYPES = [
      type("rabble", "Rabble", "vanguard", attack: 1, defence: 1, move: 3, range: 0),
      type("spearman", "Spearman", "vanguard", attack: 2, defence: 2, move: 2, range: 0, trump: [ "lighthorse" ]),
      type("elephant", "Elephant", "vanguard", attack: 4, defence: 4, move: 3, range: 0),
      type("lighthorse", "Light Horse", "cavalry", attack: 2, defence: 2, move: 3, range: 0),
      type("heavyhorse", "Heavy Horse", "cavalry", attack: 3, defence: 3, move: 2, range: 0),
      type("crossbowman", "Crossbowman", "range", attack: 2, defence: 1, move: 1, range: 2, trump: [ "elephant" ]),
      type("trebuchet", "Trebuchet", "range", attack: 1, defence: 1, move: 0, range: 3, trump: [ "dragon" ]),
      type("catapult", "Catapult", "range", attack: 3, defence: 1, move: 2, range: 3, trump: [ "dragon" ]),
      type("dragon", "Dragon", "unique", attack: 5, defence: 5, move: 10, range: 0),
      type("king", "King", "unique", attack: 2, defence: 2, move: 2, range: 0),
      type("mountain", "Mountain", "mountain", attack: 9, defence: 9, move: 0, range: 0)
    ].index_by(&:codename).freeze

    # The army by the legacy 1-based data-index. Every stored setup string
    # ("11:39|6:38|...") names units by this index, so the order is load-bearing.
    ARMY = %w[
      rabble rabble rabble
      spearman spearman
      elephant elephant
      lighthorse lighthorse
      heavyhorse heavyhorse
      crossbowman crossbowman
      trebuchet
      catapult
      dragon
      king
      mountain mountain
    ].freeze

    ARMY_SIZE = ARMY.size
    KING_INDEX = ARMY.index("king") + 1

    def self.type_at(index)
      codename = index.is_a?(Integer) && index.between?(1, ARMY_SIZE) ? ARMY[index - 1] : nil
      raise ArgumentError, "no unit at army index #{index.inspect}" unless codename

      TYPES.fetch(codename)
    end
  end
end
